### PR TITLE
Merge magnetic forward functions for prisms

### LINKED ISSUE #446: Change the shape of the magnetization in magnetic forward functions
**The problem**

Currently, the `magnetization` argument in `prism_magnetic` and the `magnetic_moments` argument in `dipole_magnetic` expect a 2d array where each row corresponds to the magnetization or magnetic moment vector for each prism or dipole, respectively.

For example, if we have 5 dipoles, the `magnetic_moments` should be:

```python
magnetic_moments = np.array(
    [
        [1.9, 2.3, 4.5],
        [0.2, -2.1, 5.7],
        [0.2, -2.1, 5.7],
        [-2.3, 7.8, 9.2],
        [0.1, 2.7, 2.1],
    ]
)
```

This introduces some inconsistencies. For example, passing arrays to the `hm.magnetic_angles_to_vec` would return a tuple of three arrays, one for each component of the magnetization/magnetic field. So in order to obtain magnetic moments from inclination, declination and amplitude, we would need to transpose those.

Moreover, the current 2d array is not in agreement how we treat coordinates in Harmonica: they are usually a tuple with `easting`, `northing` and `upward` arrays.

**The solution**

We should change the expected inputs of these functions so they take `magnetization` and `magnetic_moments` as a tuple with three 1-d arrays. Following the previous example, they should be:

```python
magnetic_moments = (
    np.array([1.9, 0.2, 0.2, -2.3 0.1]),
    np.array([2.3, -2.1, -2.1, 7.8, 2.7]),
    np.array([4.5, 5.7, 5.7, 9.2, 2.1]),
)
```

This change was discussed in [Fatiando Meeting of 2024-01-04](https://github.com/fatiando/community/blob/main/community-calls/2024.md#2024-01-04)

**Are you willing to help implement and maintain this feature?** 

Yes, but anyone that wants to tackle it is more than welcome!

We should work on this before the next release, so we don't release the new magnetic functions with an interface that will be deprecated.

### LINKED ISSUE #447: Merge magnetic forward functions into a single one
**Description of the desired feature:**

Having two functions (`prism_magnetic` and `prism_magnetic_component`) for forward modelling magnetic fields of prisms (and dipoles) is not the best. It's somewhat confusing and also it breaks the design we already have for the gravity forward functions.

We should merge them together, having a single magnetic forward function for prisms (and a single magnetic forward function for dipoles) that can compute a single component or the three components.

The new merged function should keep the name of the first one (`prism_magnetic` and `dipole_magnetic`) and should ask for a required `field` argument. This `field` argument should be able to take:

| `field` | |
|---|---|
| `b_e` | for the easting component of the magnetic vector |
| `b_n` | for the northing component of the magnetic vector |
| `b_u` | for the upward component of the magnetic vector |
| `b`   | for the three components of the magnetic vector (returned as a tuple) |

In the future, this function could be expanded to magnetic gradiometry fields:

| `field` | |
|---|---|
| `b_ee` | for the easting-easting component of the magnetic vector   |
| `b_nn` | for the northing-northing component of the magnetic vector |
| `b_uu` | for the upward-upward component of the magnetic vector     |
| `b_en` | for the easting-northing component of the magnetic vector  |
| `b_eu` | for the easting-upward component of the magnetic vector    |
| `b_nu` | for the northing-upward component of the magnetic vector   |

**Are you willing to help implement and maintain this feature?** 

Yes! But anyone wanting to tackle this is more than welcomed!

We should work on this before the next release, so we don't release the new magnetic functions with an interface that will be deprecated. Probably it would be easier to tackle this at the same time as #446

### PR DESCRIPTION
**Pull request by santisoler**
Merge the two forward functions for magnetic fields of rectangular prisms into a single one named `prism_magnetic` that includes a new `field` argument, mimicking the interface of the gravity forward functions. Make the prism_magnetic` function to ask magnetization vectors as a tuple of three arrays, where each one of these arrays. Update tests accordingly. Update the user guide page about forward modelling of prisms.

**Relevant issues/PRs:**

Fixes #446 and fixes #447

### CODE REVIEW

**Comment by santisoler** [comment]
@indiauppal, I think this is ready to be merged. Would you like to review it?

No worries if you don't, and there's no rush to do so. But I think it would be a nice experience for you!

If you agree, here is a GitHub guide on how to make a review: https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/reviewing-changes-in-pull-requests/reviewing-proposed-changes-in-a-pull-request

Let me know what do you think.

File: doc/user_guide/forward_modelling/prism.rst [hunk 1]
```diff
@@ -222,18 +221,20 @@ For example:
...
-Each row of the ``magnetization`` 2D vector corresponds to the easting,
-northing and upward components of the magnetization vector of each prism in
-``prisms``, provided in :math:`Am^{-1}`.
+The ``magnetization`` should be a tuple of three arrays: the easting, northing
+and upward components of the magnetization vector (in :math:`Am^{-1}`) for each
+prism in ``prisms``.
 
-With the :func:`harmonica.prism_magnetic` function we can compute the three
+We can use the :func:`harmonica.prism_magnetic` function to compute the three
 components of the magnetic field the prisms generates on any set of observation
```
Comment (indiauppal) [inline doc/user_guide/forward_modelling/prism.rst:236]:
```suggestion
components of the magnetic field the prisms generate on any set of observation
```

File: doc/user_guide/forward_modelling/prism.rst [hunk 2]
```diff
@@ -282,22 +283,25 @@ points.
...
+* ``"b_n"`` for the northing component, and
+* ``"b_u"`` for the upward component.
 
 .. important::
 
-   Using :func:`harmonica.prism_magnetic_component` to compute several magnetic
-   components is less efficient that using :func:`harmonica.prism_magnetic`.
-   Use the former only when a single component is needed.
+   Computing the three components at the same time with ``field="b"`` is more
+   efficient that computing each one of the three components independently.
```
Comment (indiauppal) [inline doc/user_guide/forward_modelling/prism.rst:296]:
```suggestion
   efficient than computing each one of the three components independently.
```

**Review by indiauppal — COMMENTED** [review]
A couple of tiny changes in `doc/user_guide/forward_modelling/prism.rst`, but otherwise it's ready to be merged.

**Comment by santisoler** [comment]
Thanks @indiauppal for the review! Nice catches!

I've just pushed a few more changes: I filled the docstrings of the new private functions (which were a little bit scarse), and noticed that asking for the `cast` was hard to explain, so I changed it for the `shape` of the output arrays.

If CI doesn't fail, I'll merge this right away!

> Thanks @santisoler for letting me review it!

No need to thank me, it's the other way around haha! BTW, don't feel you need to ask permission to review a PR that is interesting for you. Feel free to jump in and make the review!

**Pull request merged**